[#61511] Implement initial list timeentries command

myabc opened the pull request:
https://community.openproject.org/wp/61511

An initial patch to support listing (personal) time entries. I wrote this to "scratch an itch" - and make my own time tracking easier.

<img width="1335" alt="image" src="https://github.com/user-attachments/assets/81b5dfb7-071f-411b-8cba-0b27e4c905ef" />


Obviously full CRUD would be nice, but this is just a start.

Also, since this is the first go code I've written, please forgive if it not idiomatic!

myabc committed:
Implement initial list timeentries command

myabc committed:
Better date formatting: show weekday not year

Kharonus commented on cmd/list/time_entries.go:
```diff
@@ -0,0 +1,22 @@
+	Use:   "timeentries",
+	Short: "Lists time entries",
+	Long:  "Get a list of all personal time entries.",
+	Run:   listTimeEntry,
```
🟡 minor: `listTimeEntries`

Kharonus commented on components/resources/time_entries/filters.go:
```diff
@@ -0,0 +1,19 @@
+	User FilterOption = iota
+)
+
+func UserFilter(name string) requests.Filter {
```
🟡 I'd love to extend this filter to the - admittedly not everywhere used - filter interface. This one is used only in the work package resource, and I planned to rewrite every filter to it. So, it'd be nice to have this new filter already taking the desired form.

Kharonus replied:
I can imagine that there are some confusions :D especially as the code is half implemented around the filters. So, we can pair on it, if you want.

Kharonus commented on components/resources/time_entries/functions.go:
```diff
@@ -0,0 +1,24 @@
+
+func All() ([]*models.TimeEntry, error) {
+	var filters []requests.Filter
+	filters = append(filters, UserFilter("me"))
```
🟡 I understand that you want to fetch only the time entries assigned to `me` on a default fetch. That makes sense, but the logic for that should be in the command, not in the read function IMHO. this also gets easier, if the filter is using the "new and shiny" implementation.

Kharonus reviewed (changes requested):
Nice ❤️ thanks for your idea and effort. I'd love to put this into code after shaping it up a little bit

myabc committed:
Implement --user flag for list timeentries command

Defaults to `me`.

myabc committed:
Rework user flag to use new-style filters

Also
- validates user input (according to core app `User#login` Regexp).
- supports lists of users (separated by a comma).

myabc commented:
@Kharonus I spent a few minutes switching to the new filter implementation. Ideally, we'd also add some tests for the filters but I'm going to park this for now and let you review.

Also, lists of users should work, e.g.:
```
/openproject-cli list timeentries -u 39999,me
/openproject-cli list timeentries -u 39812,39999
```

Although the validation regexp ([copied from core](https://github.com/opf/openproject/blob/dev/app/models/user.rb#L135)) allows login names, I can't actually get this to work - neither for time entries nor work packages.:

```
/openproject-cli list timeentries -u a.coles@openproject.com
/openproject-cli list workpackages -a a.coles@openproject.com
```

I double-checked the URL escaping and that appears to be correct, so I am guessing we're using the API in an unsupported way?

Kharonus commented on components/resources/time_entries/filters/user.go:
```diff
@@ -0,0 +1,64 @@
+	"github.com/opf/openproject-cli/components/requests"
+)
+
+const validValueRegexp = `^([\pL0-9_\-@.+ ]+|[0-9]+|me)(,([\pL0-9_\-@.+ ]+|[0-9]+|me))*$`
```
❓ What is this first matcher for? the `[\pL0-9_\-@.+ ]+`

Kharonus replied:
Ah, maybe I should just read your comments first :D So, this is the regex from core itself. Maybe a comment would be nice, what this regex is supposed to match. wdyt?

myabc replied:
> ❓ What is this first matcher for? the `[\pL0-9_\-@.+ ]+`

Login / Usernames - still, as mentioned in my other comment, I'm not sure if the API *actually* supports this.

myabc replied:
> Ah, maybe I should just read your comments first :D So, this is the regex from core itself. Maybe a comment would be nice, what this regex is supposed to match. wdyt?

Good idea. I've added a comment!

Kharonus reviewed (approved):
Nice work

looks fine for me to get this merged. We might polish it a bit, once we add more filters or the other CRUD methods, but overall, job well done.

the one minor thing remains, except that :shipit:
